[14.0][IMP] Combo category behavior

ygcarvalh opened the pull request:
Implementation of an improvement aimed at adding a field that defines different behaviors for a category of a combo. At this first moment, it is possible that the category duplicates the items or that it has control of the quantity to be added for each item.

ygcarvalh committed:
[IMP] Field for combo category behavior

Needed to handle non-default behavior on the frontend.

ygcarvalh committed:
[IMP] Category behavior based on backend choice

In addition to implementing the different behavior for the category
option, it also adds error handling when adding products so that, if
there is a problem, it does not add any of the other items in the combo.

ygcarvalh committed:
[IMP] Change label for combo quantity

To avoid interpretation problems, the label of the field has been
changed.

ygcarvalh committed:
[IMP] Change in the condition of the categories of combos

To handle the implementation of the improvement targeting different
behaviors, it was necessary to change the condition to deal with the
quantities of the products of a combo.

ygcarvalh committed:
[IMP] Item duplication with quantity

Based on the amount that was defined in the backend and in the
configuration of the combo category, add the equivalent amount of
products following the definitions in the frontend.

ygcarvalh committed:
[REF] Code cleanup

luismalta commented on pos_product_template_combo/static/src/js/ProductScreen.js:
```diff
@@ -63,16 +71,73 @@ odoo.define("pos_product_template_combo.ProductScreen", function (require) {
+            _insertDuplicateOrderline(productToAdd, linesAddedToOrder) {
+                const productQuantity = productToAdd.quantity;
+                const lineNewPrice =
+                    this.selectedOrderline.get_unit_price() - productQuantity * 0.01;
+                this.selectedOrderline.set_quantity(1);
```
Acho melhor já deixar a quantidade e o preço da linha do produto original já vir correto da função _mount_products_to_add_map. Porque assim não será preciso mexer em uma linha que já foi adicionada a conta.

luismalta commented on pos_product_template_combo/static/src/js/ProductScreen.js:
```diff
@@ -63,16 +71,73 @@ odoo.define("pos_product_template_combo.ProductScreen", function (require) {
+                    this.selectedOrderline.get_unit_price() - productQuantity * 0.01;
+                this.selectedOrderline.set_quantity(1);
+                this.selectedOrderline.set_unit_price(lineNewPrice);
+                for (let i = 0; i < productQuantity; i++) {
+                    const clonedOrderline = this.selectedOrderline.clone();
+                    this.order.add_orderline(clonedOrderline);
+                    clonedOrderline.set_quantity(1);
+                    clonedOrderline.set_unit_price(0.01);
+                    linesAddedToOrder.push(clonedOrderline);
+                }
```
Como os dados dos produtos devem ser iguais para todas as duplicatas, acho melhor adicionar só uma linha com a quantidade de duplicatas correspondente. Dessa forma tira a complexidade de mais um for no algorítimo.

luismalta commented on pos_product_template_combo/static/src/js/ProductScreen.js:
```diff
@@ -63,16 +71,73 @@ odoo.define("pos_product_template_combo.ProductScreen", function (require) {
+                            productToAdd.categoryBehavior
+                        )
+                    ) {
+                        try {
+                            this._insertDuplicateOrderline(
+                                productToAdd,
+                                linesAddedToOrder
+                            );
+                        } catch (error) {
+                            this._rollbackComboOrderlines(linesAddedToOrder);
+                            this.showErrorProductComboMessage();
+                            break;
+                        }
```
Será que não conseguimos estender essa lógica do rollback para a adição de todos os produtos no combo?

ygcarvalh replied:
Não entendi sua dúvida, pode explicar? Porque até onde eu testei, é exatamente esse o comportamento que vai ser realizado por essa função do rollback.

luismalta replied:
Pelo que vi, hoje o try só cobre a adição de linhas duplicadas. A ideia seria ele cobrir também a inserção de qualquer item do combo (linha 79). Dessa forma, caso aconteça algum problema nesses casos, ele também executará o rollback.

ygcarvalh replied:
Ah, sim. Mas na linha 80 eu faço o mesmo processo que é feito nas linhas duplicadas: eu adiciono ele para o array.

luismalta replied:
Mas se der problema na adição de um item normal do combo, como essa parte do código não está no try catch, ele não executará a função do rollback.

luismalta reviewed (changes requested):
@ygcarvalh Depois consegue subir uma instância no gitlab mudando o repo / branch do combo para essa sua? Assim conseguimos testar funcionalmente também.

ygcarvalh committed:
[IMP] Item duplication for combos

A change was made to the try..catch point to handle all items in the
combo. In addition, the quantity and price of the items is in the
function that already existed for this process.

ygcarvalh commented:
@luismalta pode revisar de novo, por favor?

luismalta reviewed (approved):
LGTM. Só esperar o pipeline do Gitlab voltar a funcionar para fazer o teste funcional.
